luminoso_api/v5_client.py

```diff
@@ -0,0 +1,372 @@
+        the authentication for you.
+        """
+        self.session = session
+        self.url = ensure_trailing_slash(url)
+        # Don't warn this time; warning happened in connect()
+        self.root_url = get_root_url(url, warn=False)
+
+    def __repr__(self):
+        return '<LuminosoClient for %s>' % self.url
+
+    @classmethod
+    def connect(cls, url=None, token_file=None, token=None):
+        """
+        Returns an object that makes requests to the API, authenticated
+        with a saved or specified long-lived token, at URLs beginning with
+        `url`.
+
+        If no URL is specified, or if the specified URL is a path such as
+        '/projects' without a scheme and domain, the client will default to
+        https://analytics.luminoso.com/api/v5/.
+
+        If neither token nor token_file are specified, the client will look
+        for a token in $HOME/.luminoso/tokens.json. The file should contain
+        a single json dictionary of the format
+        `{'root_url': 'token', 'root_url2': 'token2', ...}`.
+        """
+        if url is None:
+            url = '/'
+
+        if url.startswith('http'):
+            root_url = get_root_url(url)
+        else:
+            url = URL_BASE + '/' + url.lstrip('/')
+            root_url = URL_BASE
+
+        if token is None:
+            token_file = token_file or get_token_filename()
+            with open(token_file) as tf:
+                token_dict = json.load(tf)
+            try:
+                token = token_dict[urlparse(root_url).netloc]
+            except KeyError:
+                raise LuminosoAuthError('No token stored for %s' % root_url)
+
+        session = requests.session()
+        session.auth = _TokenAuth(token)
+        return cls(session, url)
+
+    @staticmethod
+    def save_token(token, domain='analytics.luminoso.com', token_file=None):
+        """
+        Take a long-lived API token and store it to a local file.  Long-lived
+        tokens can be retrieved through the UI.  Optional arguments are the
+        domain for which the token is valid and the file in which to store the
+        token.
+        """
+        token_file = token_file or get_token_filename()
+        if os.path.exists(token_file):
+            saved_tokens = json.load(open(token_file))
+        else:
+            saved_tokens = {}
+        saved_tokens[domain] = token
+        directory, filename = os.path.split(token_file)
+        if directory and not os.path.exists(directory):
+            os.makedirs(directory)
+        with open(token_file, 'w') as f:
+            json.dump(saved_tokens, f)
+
+    def _request(self, req_type, url, **kwargs):
+        """
+        Make a request via the `requests` module. If the result has an HTTP
+        error status, convert that to a Python exception.
+        """
+        logger.debug('%s %s' % (req_type, url))
+        func = getattr(self.session, req_type)
+        result = func(url, **kwargs)
```

Comment:
I believe this line and the one above could be rewritten as `result = self.session.request(req_type, url, **kwargs)` -- a small change, but less reflection seems better.